policies/recipes/steps.go

```diff
@@ -90,8 +99,344 @@ func parsePermissions(s string) (os.FileMode, error) {
+func StepArchiveExtraction(step *osconfigpb.SoftwareRecipe_Step_ArchiveExtraction, artifacts map[string]string, stepDir string) error {
+	filename, ok := artifacts[step.ArchiveExtraction.GetArtifactId()]
+	if !ok {
+		return fmt.Errorf("%q not found in artifact map", step.ArchiveExtraction.GetArtifactId())
+	}
+	switch step.ArchiveExtraction.GetType() {
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_ZIP:
+		return extractZip(filename, step.ArchiveExtraction.Destination)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR:
+		return extractTar(filename, step.ArchiveExtraction.Destination)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_GZIP:
+		compressed, err := os.Open(filename)
+		if err != nil {
+			return err
+		}
+		defer compressed.Close()
+		reader, err := gzip.NewReader(compressed)
+		if err != nil {
+			return err
+		}
+		defer reader.Close()
+		decompressed, err := ioutil.TempFile(stepDir, "archive-*.tar")
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(decompressed, reader)
+		decompressed.Close()
+		if err != nil {
+			return err
+		}
+		return extractTar(decompressed.Name(), step.ArchiveExtraction.Destination)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_BZIP:
+		compressed, err := os.Open(filename)
+		if err != nil {
+			return err
+		}
+		defer compressed.Close()
+		reader := bzip2.NewReader(compressed)
+		if err != nil {
+			return err
+		}
+		decompressed, err := os.Create(filepath.Join(stepDir, "archive.tar"))
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(decompressed, reader)
+		decompressed.Close()
+		if err != nil {
+			return err
+		}
+		return extractTar(decompressed.Name(), step.ArchiveExtraction.Destination)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_LZMA:
+		compressed, err := os.Open(filename)
+		if err != nil {
+			return err
+		}
+		defer compressed.Close()
+		reader, err := lzma.NewReader2(compressed)
+		if err != nil {
+			return err
+		}
+		decompressed, err := os.Create(filepath.Join(stepDir, "archive.tar"))
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(decompressed, reader)
+		decompressed.Close()
+		if err != nil {
+			return err
+		}
+		return extractTar(decompressed.Name(), step.ArchiveExtraction.Destination)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_XZ:
+		compressed, err := os.Open(filename)
+		if err != nil {
+			return err
+		}
+		defer compressed.Close()
+		reader, err := xz.NewReader(compressed)
+		if err != nil {
+			return err
+		}
+		decompressed, err := os.Create(filepath.Join(stepDir, "archive.tar"))
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(decompressed, reader)
+		decompressed.Close()
+		if err != nil {
+			return err
+		}
+		return extractTar(decompressed.Name(), step.ArchiveExtraction.Destination)
+	default:
+		return fmt.Errorf("Unrecognized archive type %q", step.ArchiveExtraction.GetType())
+	}
+}
+
+func zipIsDir(name string) bool {
+	return strings.HasSuffix(name, "/")
```

Comment:
You may need to also check os.PathSeparator

Reply:
The zip format seems to specify forward slashes specifically, are files not to spec common? From looking into it historically they seem to be reasonably common. I quickly added code in to deal with different slash directions but I'm curious if that's necessary.

Reply:
Test it on Windows to be sure, if it's still '/' you can leave it

Reply:
the zip spec strictly indicates forward slash, i don't think it's a platform variance so much as checking whether or not all zip producing tools follow the spec. if these zip files are likely to occur then yeah we should have supporting logic

Reply:
Essentially what I did was convert '/' to os.PathSeparator then for directories just checked if the path ends with os.PathSeparator. This means on Windows either '/' or '\' will work as path separators but on Linux only '/' will work. This means that zips that use '\' will create directories on windows but on Linux will extract according to the spec.

Reply:
i think your comment had some misformatting. but why would we want this differentiated outcome? why not use the simpler 'any trailing slash means a dir' logic? we're adding complex edge cases, and i don't see the motivation

Reply:
It would take more than just checking for dirs like that to actually support '\' as a path separator on Linux. Essentially I would prefer to be consistent about whether '\' was treated as a path separator on each distro.

For Windows some historical tools have created zips with '\' as a separator (and my impression is that modern tools will often handle unzipping them correctly) and also go's os library itself will treat '\' as a path separator so not having '\' be a path separator at all would be fairly difficult.

Linux doesn't have that historical baggage (and as far as I'm aware the current Linux unzip utility doesn't treat '\' as a path separator) and treating '\' as a separator would break handling of zips where '\' is used in a file name and isn't a separator.

I don't have great information on whether finding zips not to spec or filenames with '\' in them is more likely but I erred on the side of sticking to the spec and working the same as the Linux unzip. If this explanation isn't clear we can talk about it more in chat.

Reply:
you seem to be commenting `" as a path separator`